### PR TITLE
c8d/import: Don't gc unnamed images

### LINKED ISSUE #45416: containerd backend: loading images (from tar) w/o refs discards everything
### Description

If you have an image tar, but in that tar there are no annotations to specify what the image ref should be no image object is ever created and as such containerd's GC cleans up all the content once the import operation is completed.

This is a change in behavior from the graphdriver backend.

### Reproduce

With the containerd storage backend:

```
docker buildx create --use
docker buildx build --output=type=docker,dest=foo.tar -<<<"FROM busybox"
docker load -i foo.tar
```

The `docker load` does effectively nothing, does not output anything either.
This is because `client.Import` from the containerd client does not create images for images without an image ref annotation in the descriptors.

### Expected behavior

This should create a dangling image ref.

### docker version

```bash
Client:
 Version:           23.0.2+azure-1
 API version:       1.42
 Go version:        go1.19.6
 Git commit:        569dd73db13099a7c3104d73aa15117b359045bc
 Built:             Mon Mar 27 15:44:26 UTC 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          dev
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.3
  Git commit:       489543cd25
  Built:            Wed Apr 26 17:57:17 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.0+azure-1
  GitCommit:        1fbd70374134b891f97ce19c70b6e50c7b9f4e0d
 runc:
  Version:          1.1.5
  GitCommit:        f19387a6bec4944c770f7668ab51c4348d9c2f38
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.10.4+azure-1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx

Server:
 Containers: 4
  Running: 1
  Paused: 0
  Stopped: 3
 Images: 198
 Server Version: dev
 Storage Driver: overlayfs
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: local
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1fbd70374134b891f97ce19c70b6e50c7b9f4e0d
 runc version: f19387a6bec4944c770f7668ab51c4348d9c2f38
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-1036-azure
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.63GiB
 Name: dev2
 ID: ccb953ef-cff6-4d73-94de-24e41bbff5c2
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 28
  Goroutines: 41
  System Time: 2023-04-26T23:12:51.832041921Z
  EventsListeners: 0
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by vvoland**
- Fixes: https://github.com/moby/moby/issues/45416

Create dangling images for imported images which don't have a name annotation attached. Previously the content got loaded, but no image referencing it was created which caused it to be garbage collected immediately.

**- What I did**

**- How I did it**

**- How to verify it**
```bash
$ docker buildx create --use
$ docker buildx build --output=type=docker,dest=foo.tar -<<<"FROM busybox"
...
$ docker load -i foo.tar
Loaded image ID: sha256:9efda7ee71e76fc5a755b9e3e6f065544380306e6bd77eb20d44772bc61d13ab
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/image_exporter.go [hunk 1]
```diff
@@ -123,8 +136,13 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outSt
 		}
 
 		name := img.Name
-		if named, err := reference.ParseNormalizedNamed(img.Name); err == nil {
-			name = reference.FamiliarName(named)
+		loadedMsg := "Loaded image"
+
+		if isDanglingImage(img) {
+			name = img.Target.Digest.String()
+			loadedMsg = "Loaded image ID"
```
Comment (vvoland) [inline daemon/containerd/image_exporter.go:143]: The conditional message is what the non-c8d Docker does.
Wondering if it's crucial to keep the output compatible? Maybe we could just have one "Loaded image: <image>" for both cases?

Comment (thaJeztah) [reply daemon/containerd/image_exporter.go:143]: It looks like there's at least some code checking for this; https://grep.app/search?q=Loaded%20image%20ID

I agree that we should look into some of these; perhaps it should show _more_ details (if multiple architectures are loaded etc etc), and close with a `Loaded image` (or something).

if we're gonna "break" output, we should probably do some of that with a big bang; related to that;

- https://github.com/moby/moby/issues/41838

**Comment by thaJeztah** [comment]
CI failures are unrelated and can be ignored; see

- https://github.com/moby/moby/pull/45415

File: daemon/containerd/image_exporter.go [hunk 2]
```diff
@@ -103,10 +103,23 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
 // complement of ExportImage.  The input stream is an uncompressed tar
 // ball containing images and metadata.
 func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error {
-	// TODO(vvoland): Allow user to pass platform
-	platform := cplatforms.All
-	imgs, err := i.client.Import(ctx, inTar, containerd.WithImportPlatform(platform))
+	opts := []containerd.ImportOpt{
+		// TODO(vvoland): Allow user to pass platform
+		containerd.WithImportPlatform(cplatforms.All),
```
Comment (thaJeztah) [inline daemon/containerd/image_exporter.go:108]: Can you create a tracking-ticket for this?

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@cpuguy83 PTAL

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**